Add migration to fix incorrect product review count.

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

In https://github.com/woocommerce/woocommerce/pull/28624 we fixed the root cause for $27688, however, we still needed a migration to fix the product review counts. This PR adds that migration, to fix the counts, we loop over all meta props with keys `_wc_review_count` and update them if they are incorrect and post belongs to a product.

Closes #27688.

### How to test the changes in this Pull Request:

1. Run this query to invalidate all product review count in your shop:
```sql
UPDATE %postmeta_table_name%
SET meta_value = '9999'
WHERE meta_key = '_wc_review_count';
```
2. Go to any product page in the frontend and check that the review count is incorrect.
3. Open a WP shell and run this command:
```php
include_once dirname( WC_PLUGIN_FILE ) . '/includes/wc-update-functions.php';
WC_Install::run_update_callback('wc_update_500_fix_product_review_count'); // Keep on calling this till output is false. 
```
4. Go back to the product page in the frontend, refresh and check that the review count is now correct.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Add migration to fix existing product review count.

vedanshujain committed:
Add utility function for getting review counts in bulk for convinience.

Also refactors get_review_count_for_product to use the new method so that we only would have to maintain one method.

vedanshujain committed:
Add unit tests for review count methods.

vedanshujain committed:
Add migration function for fixing existing product review counts.

roykho commented on includes/class-wc-comments.php:
```diff
@@ -340,6 +340,46 @@ public static function get_average_rating_for_product( &$product ) {
+		$product_id_string = implode( "','", array_map( 'esc_sql', $product_ids ) );
+
+		$review_counts = $wpdb->get_results(
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
```
How come we're bypassing `prepare` here?

vedanshujain replied:
This is because of `$product_id_string` which we were expanding for `IN` clause. I thought this is the standard way to use when doing interpolation for `IN` clause.

But after your comment, I looked in detail and see how WP core is doing it, and found this reference: https://github.com/pantheon-systems/wordpress-network/blob/master/wp-includes/class-wp-meta-query.php#L639. 
This approach still ignores phpcs, but does not bypass `prepare` so I have changed the code to use this approach in https://github.com/woocommerce/woocommerce/pull/28814/commits/9f9475a18b6a80f1b256041b4c9e56d40a45ebe8. Let me know if this makes sense.

roykho reviewed (commented):
Tested well for me. I had to checkout release/4.8 and followed the steps in the original issue to reproduce it. Then I checkout this PR and followed the testing steps and it fixed the arbitrary review count `9999` to the correct count in the product frontend.

Thanks for including the test cases. I just left small comments.

vedanshujain committed:
Addressed PR feedback.

1: Use placeholder to be able to use wpdb->prepare for IN query.

Update version number.

vedanshujain commented on includes/wc-update-functions.php:
```diff
@@ -2219,3 +2219,67 @@ function wc_update_450_sanitize_coupons_code() {
+}
+
+/**
+ * Update DB version to 4.5.0.
```
Good catch, addressed in https://github.com/woocommerce/woocommerce/pull/28814/commits/9f9475a18b6a80f1b256041b4c9e56d40a45ebe8

roykho reviewed (approved):
LGTM!
